### PR TITLE
Add Portis web3 provider via a wrapper

### PR DESCRIPTION
**Pull request by jessgusclark**
Adds support for Portis as a web3 provider.

It also adds a loading wheel after the user selects a provider. This is also shown when the user is connecting the DataVault. i.e. User clicks on Metamask, the waiting wheel is shown until the user is connected.

## Details

Portis is not an EIP1193 provider, so after Portis is passed to rLogin, a wrapper is created that allows EIP1193 methods to be called using `request`. The tested methods are:

- `eth_chainId`
- `eth_accounts`
- `net_version`
- `personal_sign` - Portis requires that the message be encoded as HEX, which this wrapper does for the user. I'm unsure if this is "standard" (before 1193) or not, so right now this wrapper will only be used with Portis.
- `eth_sendTransaction`

## Apps tested with:

- [x] RIF Id Manager
- [x] Email VC Issuer
- [x] Sample App with all 3 variants: `dapp`, `open`, `permissioned`

### CODE REVIEW

File: src/Core.tsx [hunk 1]
```diff
@@ -280,6 +283,11 @@ export class Core extends React.Component<IModalProps, IModalState> {
     // send disconnect method to wallet connect
     this.disconnectWC(provider)
 
+    // portis
+    if (provider.isPortis) {
+      provider._portis.logout()
```
Comment (ilanolkies) [inline src/Core.tsx:288]: We should await this, it is a `Promise` - https://github.com/portis-project/web-sdk/blob/develop/packages/portis-web3/src/index.ts#L126

should we move this to a `disconnectPortis` method? As for WC

File: src/lib/portisWrapper.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+          { method, params },
+          (err: string | null, res: { result?: string, error?: Error }) =>
+            err ? reject(res.error) : resolve(res.result)
+        ))
+      }
+    }
+  }
+}
+
+export const portisWrapper = (provider: any) => new Proxy(provider, handler)
```
Comment (ilanolkies) [inline src/lib/portisWrapper.ts:33]: Excellent! Some minor things:
- Should we add support for `eth_signTypedData` and `eth_sendTransaction` too? I don’t know if they work already
- Can we list the methods we tested?
- Let’s add some comments about this code

```suggestion
// This wrapper makes Portis Web SDK EIP-1193 compatible
export const portisWrapper = (provider: any) => new Proxy(provider, handler)
```

File: src/Core.tsx (hunk 1 above)
Comment (ilanolkies) [reply src/Core.tsx:288]: I see now WC `disconnect` returns a `Promise` too
https://github.com/WalletConnect/walletconnect-monorepo/blob/v2.0/packages/ethereum-provider/src/index.ts#L116

**Review by ilanolkies — COMMENTED** [review]
Nice onee!

File: src/Core.tsx [hunk 3]
```diff
@@ -316,11 +324,12 @@ export class Core extends React.Component<IModalProps, IModalState> {
       setLightboxRef={this.setLightboxRef}
       mainModalCard={this.mainModalCard}
     >
-      {currentStep === 'Step1' && <WalletProviders userProviders={userProviders} />}
+      {currentStep === 'Step1' && <WalletProviders userProviders={userProviders} setLoading={() => this.setState({ currentStep: 'loading' })} />}
       {currentStep === 'Step2' && <SelectiveDisclosure sdr={sdr!} backendUrl={backendUrl!} fetchSelectiveDisclosureRequest={this.fetchSelectiveDisclosureRequest} onConfirm={this.onConfirmSelectiveDisclosure} />}
       {currentStep === 'Step3' && <ConfirmSelectiveDisclosure did={(chainId && address) ? did : ''} sd={sd!} onConfirm={this.onConfirmAuth} />}
       {currentStep === 'error' && <ErrorMessage title={errorReason?.title} description={errorReason?.description}/>}
       {currentStep === 'wrongNetwork' && <WrongNetworkComponent supportedNetworks={supportedChains} isMetamask={isMetamask(provider)} changeNetwork={this.changeMetamaskNetwork} />}
+      {currentStep === 'loading' && <Loading size={10} color="#008FF7" />}
```
Comment (ilanolkies) [inline src/Core.tsx:332]: Cool

File: src/Core.tsx (hunk 1 above)
Comment (jessgusclark) [reply src/Core.tsx:288]: Great! I have added a `disconnect` method to the wrapper. This mirrors how walletconnect works. So, now when disconnecting, I can check if this method exists and call it.

File: src/lib/portisWrapper.ts (hunk 2 above)
Comment (jessgusclark) [reply src/lib/portisWrapper.ts:39]: I have added additional comments to the wrapper and added a section in the readme for Portis. Regarding `eth_signTypeData` it does not work out of the box as Portis' implementation is a bit different than Metamask's. Should I take the time now to implement this?

**Comment by jessgusclark** [comment]
Rebased.

File: src/lib/portisWrapper.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+              { method: tMethod, params: tPrams },
+              (err: string | null, res: { result?: string, error?: Error }) =>
+                err ? reject(res.error) : resolve(res.result))
+
+            // for sendTransaction, get the gasPrice from the last block if the developer did not
+            // include it. Portis does not calculate it correctly, causing the transaction to fail
+            if ((method === 'eth_sendTransaction' || method === 'eth_sendRawTransaction') && !params[0].gasPrice) {
+              return target.send({ method: 'eth_getBlockByNumber', params: ['latest', false] },
+                (err: string | null, res: any) =>
+                  err ? reject(res.err) : sendTransaction(method, [{ ...params[0], gasPrice: res.result.minimumGasPrice }])
```
Comment (ilanolkies) [inline src/lib/portisWrapper.ts:38]:
```suggestion
                  err ? reject(res.err) : sendTransaction(method, [{ ...params[0], gasPrice: res.result.minimumGasPrice * 1.01 }])
```
Or we can set even a little more

> Nodes that forward transactions could check that the advertised gas price in a transaction is at least 1% higher than the minimum. This assures the transaction a lifetime of 100 blocks assuming a constantly increasing block minGasPrice.  Ref: https://github.com/rsksmart/RSKIPs/blob/master/IPs/RSKIP09.md

**Review by ilanolkies — CHANGES_REQUESTED** [review]

**Comment by ilanolkies** [comment]
Rebased version bumping

**Comment by ilanolkies** [comment]
Let's add the last comment and merge the beta version into develop. After we give it a test we can bump and release major.

File: src/lib/portisWrapper.ts (hunk 4 above)
Comment (jessgusclark) [reply src/lib/portisWrapper.ts]: Added.

**Comment by jessgusclark** [comment]
After talking with @miltonwg , I have updated the description in the Readme to indicate that the wrapper will add the `gasPrice` when sending `request`. However, the developer will need to add a `gasPrice` if they use `send` or `sendAsync`.

I also bumped the `gasPrice` by 1% as requested by @ilanolkies . This PR is ready to go, below is the status of the App's support of Portis. I checked the Apps below using `yarn link` as a few extra commits have been added.

- [x] VC Issuer - Worked fine as it does not have any transactions, only signatures. Portis support is already merged into main. [deployed to production here](https://email-verifier.identity.rifos.org/)
- [x] RIF Id Manager - As Ken found, the identity section does not work. After talking with Milton, it was decided to show a message letting the user know they can't update their data. [A PR is ready to be merged here](https://github.com/rsksmart/rif-identity-manager/pull/81). 
- [x] RNS Manager React - Works locally, however, the published rLogin package 1.0.7-beta.1 does not support rawTransactions, so [this PR](https://github.com/rnsdomains/rns-manager-react/pull/417) will need to wait until the next version is released then updated.

---
update: this comment was originally posted on the RIF Identity Manager PR.

**Comment by ilanolkies** [comment]
Do we add it to v1.0.8? In that case let;s merge it before #96

**Review by ilanolkies — APPROVED** [review]

**Pull request merged**